cachito/workers/pkg_managers/gomod.py

```diff
@@ -472,6 +459,81 @@ def go_list_deps(pattern: Literal["./...", "all"]) -> Iterator[GoPackage]:
         }
 
 
+def _parse_local_modules(
+    modules_json_stream: str, app_source_path: Path, git_dir_path: Path, request: dict[str, Any]
+) -> tuple[dict[str, str], list[GoModule]]:
+    """Parse the data regarding the main module and all existing workspaces.
+
+    This function takes the data returned by the command "go list -m -json", which returns
+    the current go module being processed, or a list of all existing workspaces listed in
+    the "go.work" file.
+
+    The main module's version is determined by evaluating the tags applied to the git repo.
+
+    :param modules_json_stream: the stream returned by "go list -m json"
+    :param app_source_path: the full path to the main go module
+    :param git_dir_path: the full path to the git repository
+    :param request: the current Cachito request
+    :return: a tuple containing the main module and a list of workspaces
+    """
+    main_module, workspace_modules = _parse_modules_json_stream(
+        app_source_path, modules_json_stream
+    )
+
+    main_module_name = main_module.path
+
+    main_module_version = get_golang_version(
+        main_module_name,
+        git_dir_path,
+        request["ref"],
+        update_tags=True,
+        subpath=(
+            None
+            if app_source_path == git_dir_path
+            else str(app_source_path).replace(f"{git_dir_path}/", "")
+        ),
+    )
+
+    for module in workspace_modules:
+        module.version = main_module_version
```

Comment:
As far as I could tell, Cachito is ignoring version tags applied to submodules. So we can just use the main module version for all workspaces, since we're pretty much keeping the current behavior.

We should probably fix this in a follow up PR, though.

Reply:
I tested it with:
```
{
  "repo": "https://github.com/cachito-testing/gomod-pandemonium",
  "ref": "0c6890c3280a00271891f4bd04705a56151428f0",
  "pkg_managers": ["gomod"]
}
```

And I got:
```
{
    "name": "github.com/cachito-testing/gomod-pandemonium/terminaltor",
    "purl": "pkg:golang/github.com%2Fcachito-testing%2Fgomod-pandemonium%2Fterminaltor@v0.1.0",
    "type": "library",
    "version": "v0.1.0"
}
```

Although the version for terminaltor should've been [v1.0.0](https://github.com/cachito-testing/gomod-pandemonium/releases/tag/terminaltor%2Fv1.0.0).

Reply:
(The SBOM also had an entry in which the version for terminaltor was `./terminaltor`, so the behavior is a little odd)

Reply:
I think Cachito does support submodule versioning. But only if you tell it to process the submodule :smiling_face_with_tear:

Reply:
So, I tried specifying all modules using the input you provided, and it still is correctly picking up for terminaltor:
```
  {
    "name": "github.com/cachito-testing/gomod-pandemonium/terminaltor",
    "purl": "pkg:golang/github.com%2Fcachito-testing%2Fgomod-pandemonium%2Fterminaltor@v1.0.0",
    "type": "library",
    "version": "v1.0.0"
  }
```

Which means we're still preserving the same behavior: submodule versions will only be picked up if the submodule is specified as a Cachito input package.

Reply:
I didn't see any duplication in the resulting SBOM that happens as a side effect of specifying all the submodules/workspaces present in the repo. The terminaltor module has a duplication regardless of what input packages you specify, though:

```
 {
    "name": "github.com/cachito-testing/gomod-pandemonium/terminaltor",
    "purl": "pkg:golang/github.com%2Fcachito-testing%2Fgomod-pandemonium%2Fterminaltor@v1.0.0",
    "type": "library",
    "version": "./terminaltor"
  },
  {
    "name": "github.com/cachito-testing/gomod-pandemonium/terminaltor",
    "purl": "pkg:golang/github.com%2Fcachito-testing%2Fgomod-pandemonium%2Fterminaltor@v1.0.0",
    "type": "library",
    "version": "v1.0.0"
  }
```

This does not happen with `weird`... I'm not sure what is causing it yet. :thinking:

Reply:
On the one hand, with workspaces it doesn't make any sense to require that the user specify all submodules

On the other hand, ¯\\\_(ツ)\_/¯

Reply:
Well, so I found out that the previous solution was indeed creating duplicate results in the SBOM. I rewrote it so:
- it picks the right version for each workspace
- the user does not need to specify multiple input packages anymore